Sources/GRPC/HTTP2ToRawGRPCStateMachine.swift

```diff
@@ -0,0 +1,1277 @@
+    switch self {
+    case .requestIdleResponseIdle:
+      preconditionFailure("Invalid state: the request stream is still closed")
+
+    case .requestOpenResponseIdle,
+         .requestClosedResponseIdle:
+      let error = GRPCError.InvalidState("Response headers must be sent before response message")
+      return .completePromise(promise, .failure(error))
+
+    case let .requestOpenResponseOpen(state):
+      let stateAndAction = state.send(
+        buffer: buffer,
+        allocator: allocator,
+        compress: compress,
+        promise: promise
+      )
+      self = stateAndAction.state
+      return stateAndAction.action
+
+    case let .requestClosedResponseOpen(state):
+      let stateAndAction = state.send(
+        buffer: buffer,
+        allocator: allocator,
+        compress: compress,
+        promise: promise
+      )
+      self = stateAndAction.state
+      return stateAndAction.action
+
+    case .requestClosedResponseClosed:
+      return .completePromise(promise, .failure(GRPCError.AlreadyComplete()))
+
+    case ._modifying:
+      preconditionFailure("Left in modifying state")
+    }
+  }
+
+  mutating func send(
+    status: GRPCStatus,
+    trailers: HPACKHeaders,
+    promise: EventLoopPromise<Void>?
+  ) -> HTTP2ToRawGRPCStateMachine.Action {
+    switch self {
+    case .requestIdleResponseIdle:
+      preconditionFailure("Invalid state: the request stream is still closed")
+
+    case let .requestOpenResponseIdle(state):
+      let stateAndAction = state.send(status: status, trailers: trailers, promise: promise)
+      self = stateAndAction.state
+      return stateAndAction.action
+
+    case let .requestClosedResponseIdle(state):
+      let stateAndAction = state.send(status: status, trailers: trailers, promise: promise)
+      self = stateAndAction.state
+      return stateAndAction.action
+
+    case let .requestOpenResponseOpen(state):
+      let stateAndAction = state.send(status: status, trailers: trailers, promise: promise)
+      self = stateAndAction.state
+      return stateAndAction.action
+
+    case let .requestClosedResponseOpen(state):
+      let stateAndAction = state.send(status: status, trailers: trailers, promise: promise)
+      self = stateAndAction.state
+      return stateAndAction.action
+
+    case .requestClosedResponseClosed:
+      return .completePromise(promise, .failure(GRPCError.AlreadyComplete()))
+
+    case ._modifying:
+      preconditionFailure("Left in modifying state")
+    }
+  }
+}
+
+// MARK: - Helpers
+
+extension HTTP2ToRawGRPCStateMachine {
+  static func makeResponseHeaders(
+    contentType: ContentType,
+    responseEncoding: String?,
+    acceptableRequestEncoding: String?,
+    userProvidedHeaders: HPACKHeaders,
+    normalizeUserProvidedHeaders: Bool
+  ) -> HPACKHeaders {
+    let capacity = 2 // :status, content-type
+      + (responseEncoding == nil ? 0 : 1)
+      + (acceptableRequestEncoding == nil ? 0 : 1)
+      + userProvidedHeaders.count
```

Comment:
Is the math worthwhile here, or should we just do 4 + count? Seems like excessive branching for minimal savings.

Reply:
You're probably right tbh.